internal/encoding/decoder.go

```diff
@@ -0,0 +1,58 @@
+package encoding
+
+import (
+	"sync"
+)
+
+// Decoder decodes the contents of b into a v representation.
+// It's primarily used for decoding contents of a file into a map[string]interface{}.
+type Decoder interface {
+	Decode(b []byte, v interface{}) error
+}
+
+const (
+	ErrDecoderNotFound                = encodingError("decoder not found for this format")
+	ErrDecoderFormatAlreadyRegistered = encodingError("decoder already registered for this format")
+)
+
+// DecoderRegistry can choose an appropriate Decoder based on the provided format.
+type DecoderRegistry struct {
+	decoders map[string]Decoder
+
+	mu sync.RWMutex
```

Comment:
Not sure if matters, but can use sync.Map to avoid explicitly synchronize access to the map.

Reply:
Hm, yeah, interesting idea. For now, I think type safety is more important than performance, especially since this mutex won't be locked more than 5-10 times during an application's lifecycle. But I will certainly play with it in the next round of changes (this is not the final form of the encoding package).

Thanks!